### DISCUSSION TITLE
Calculating average number of intersections per road distance unit

### DISCUSSION THREAD

**Discussion by wipfli** [onthegomap/planetiler#1071] [Q&A]
I would like to apply different styling for roads in cities and roads on the country side, and this for the same OSM `highway` tag value.

One idea for a differentiator between urban and rural was to count how many intersections a road has per unit length.

Is there a way to calculate this statistic during OSM parsing and attach the outputs to the polyline source features? It could also be interesting to count traffic signals...

**Comment by msbarry** [comment]
There are profile hooks you can override to process OSM nodes, ways, and relations during pass 1. You could store node ids for stop signs/lights or even every node that's part of a road. Then on pass 2 process you could count how many nodes are also in those sets (or more than 1 road). That wouldn't  be exact because it would miss roads that cross over or under each other and don't connect but seems like it would be a good enough proxy?

**Reply by msbarry** [reply]
Openmaptiles profile uses these hooks to determine the class of a highway exit marker based on the classes of the roads it's a part of.

**Reply by wipfli** [reply]
Ah that's a great idea. So basically in pass 1 I would do something like

```java
  private final LongSet trafficSignalNodes = new LongHashSet();
  // ...
  @Override
  public void preprocessOsmNode(OsmElement.Node node) {
    if (node.hasTag("highway", "traffic_signals")) {
      synchronized (trafficSignalNodes) {
        trafficSignalNodes.add(node.id());
      }
    }
  }
```

and then in pass 2 I would do maybe

```java
void process(SourceFeature sourceFeature, FeatureCollector features) {
  // compute length of sourceFeature
  // check how many nodes of sourceFeature are in the trafficSignalNodes set
  // compute linear density
}
```
is that roughly how I could do it?

**Reply by msbarry** [reply]
Yes, something like that should work. There are other variations if either the set gets too big in memory or if synchronization slows things down too much but I'd try your version first.

**Reply by wipfli** [reply]
Playing around with these options if found that actually a good signal for urban vs non-urban `highway=primary` is the presence of a `ref` tag. In cities there are usually no ref tags on primary roads...

**Comment by adamfranco** [comment]
Not related to Planetiler, but I tried to resolve a similar problem in my Curvature program: https://roadcurvature.com/2020/10/25/avoiding-congestion-and-conflict-zones/

`traffic_calming=*`, `junction=roundabout|circular`, and `parking:lanes=*` in addition to `highway=stop|traffic_signals|give_way` all make useful indicators of more urban environments.
